src/core/storage/sframe_data/groupby_aggregate_impl.cpp

```diff
@@ -347,46 +395,140 @@ void group_aggregate_container::flush_segment(size_t segmentid) {
     delete hash_entries.second;
   }
 
-  for (auto& item: local_sorted) {
-    for (auto& value: item.values) {
+  for (auto& item : local_sorted) {
+    for (auto& value : item.values) {
       value->partial_finalize();
     }
   }
   // ok. now we can write! lock the file
-  std::unique_lock<turi::mutex> filelock(segments[segmentid].file_lock);
-  oarchive oarc;
-  for (auto& item: local_sorted) {
-    oarc << item;
-    // write into the iterator
-    *(segments[segmentid].outiter) = std::string(oarc.buf, oarc.off);
-    ++(segments[segmentid].outiter);
-    oarc.off = 0;
+  size_t round_robin = (++merry_go_round_) % local_buffer_set_.size();
+
+  auto& local_buffer = local_buffer_set_[round_robin];
+
+  local_buffer.refctr_++;
+  /// lock the segmeent from local_buffer before using it
+  {
+    std::lock_guard<turi::simple_spinlock> slk(
+        local_buffer.sa_seg_locks_[segmentid]);
+
+    logstream(LOG_INFO) << "flush buffer from task_id: " << tss_.id_
+                            << ", segment_id: "<< segmentid << ", on buffer: " << round_robin << std::endl;
+
+    auto outiter = local_buffer.sa_buffer_ptr_->get_output_iterator(segmentid);
+    oarchive oarc;
+    for (auto& item : local_sorted) {
+      oarc << item;
+      // write into the iterator
+      *(outiter) = std::string(oarc.buf, oarc.off);
+      ++(outiter);
+      oarc.off = 0;
+    }
+
+    free(oarc.buf);
```

Comment:
Why doesn't the destructor of oarc free this when it goes out of scope on the next line?  It seems we shouldn't have to do this and it has potential to cause a double-free.

Reply:
the destructor of oarchive is empty. that's why I need to manually destroy it.

Reply:
Interesting.  That seems like a bug.

Reply:
Looking at the code in oarchive, buf is typically a pointer into the std::vector<char> that is also contained within.  So it shouldn't be freed typically.

Reply:
However, I see now that this is just keeping the old code.  Not sure I like it. 

Let's see if this works: 

std::vector<char> buffer; 

for(auto& item : local_sorted) { 
   buffer.clear();
   oarchive(buffer) << item;
   *outiter = std::string(buffer.data(), buffer.size());
   ++outiter;
}

Reply:
```
 std::vector<char> buf_;
 oarchive oarc(buf_);
    for (auto& item : local_sorted) {
      oarc << item;
      // write into the iterator
      *(outiter) = std::string(oarc.buf, oarc.off);
      ++(outiter);
      oarc.off = 0;
    }
```

Reply:
`operator<<` doesn't support rvalue. so let's get it in a little bit more verbose. 

```
    std::vector<char> buf_;
    for (auto& item : local_sorted) {
      buf_.clear();
      // hopefully, compiler can optimize this
      oarchive out(buf_);
      out << item;
      *(outiter) = std::string(buf_.data(), buf_.size());
      ++(outiter);
    }
```